components/board/numato/board.go

```diff
@@ -346,9 +355,17 @@ func (b *numatoBoard) WriteAnalog(ctx context.Context, pin string, value int32,
 
 func (b *numatoBoard) Close(ctx context.Context) error {
 	atomic.AddInt32(&b.closed, 1)
+
+	// Send a serial command so the goroutine doesn't get stuck trying to read a string after the port is closed.
```

Comment:
I think we need a much more detailed comment here. It took us hours to figure this out, and the nuance is not contained in this line. It should be something more like the following:

> The coroutine is blocked in the call to `in.ReadString`, and there is no way to interrupt that by cancelling a context or other signals. We also cannot move the `in.ReadString` call into `doSendReceive` directly, because then there is no way for it to time out if the board is not set up properly. Instead, we need the coroutine to complete that call so it wakes up and sees it should exit. Although closing the device port will accomplish this on certain OSes, on Mac you cannot close a device that is currently being read from, resulting in a deadlock (we can't close it until the coroutine stops reading, and the coroutine won't stop reading if it's not closed). So instead, we send the Numato board a command. The response back will wake up the coroutine, which will then notice it's time to exit, and we can close the device afterwards.

(that was too wordy. There's probably a good compromise in between your version and mine)